src/combinatorics/simple_graph/ends/defs.lean

```diff
@@ -0,0 +1,280 @@
+end
+
+/--
+Assuming `G` is preconnected and `K` not empty, given any connected component `C` outside of `K`,
+there exists a vertex `k ∈ K` adjacent to a vertex `v ∈ C`.
+-/
+lemma adj [Gc : G.preconnected] (hK : K.nonempty) :
+  ∀ (C : G.comp_out K), ∃ (ck : V × V), ck.1 ∈ C ∧ ck.2 ∈ K ∧ G.adj ck.1 ck.2 :=
+begin
+  refine connected_component.ind (λ v, _),
+  let C : G.comp_out K := comp_out.of_vertex G v.prop,
+  let dis := set.disjoint_iff.mp C.outside,
+  by_contra' h,
+  suffices : set.univ = (C : set V),
+  { exact dis ⟨hK.some_spec, this ▸ (set.mem_univ hK.some)⟩, },
+  symmetry,
+  rw set.eq_univ_iff_forall,
+  rintro u,
+  by_contradiction unC,
+  obtain ⟨p⟩ := Gc v u,
+  obtain ⟨x, y, xy, xC, ynC⟩ :=
+    p.disagreeing_adj_pair (C : set V) (comp_out.of_vertex_mem G v.prop) unC,
+  refine @nonadj V G K C _,
+  have : (G.out K).connected_component_mk v = comp_out.of_vertex G v.prop, by
+    simp only [connected_component.eq, subtype.coe_eta],
+  rw this at h,
+  exact ⟨x, y, xC, ynC, λ (yK : y ∈ K), h ⟨x, y⟩ xC yK xy, xy⟩,
+end
+
+/--
+If `K ⊆ L`, the components outside of `L` are all contained in a single component outside of `K`.
+-/
+@[reducible] def hom (C : G.comp_out L) (h : K ⊆ L) : G.comp_out K := C.map (G.out_hom h)
+
+lemma sub_hom (C : G.comp_out L) (h : K ⊆ L) : (C : set V) ⊆ (C.hom h : set V) :=
+begin
+  rintro c cC,
+  simp only [set.mem_compl_iff, set_like.mem_coe, mem_supp_iff] at cC ⊢,
+  obtain ⟨cL, rfl⟩ := cC,
+  exact ⟨λ h', cL (h h'), rfl⟩,
+end
+
+lemma hom_eq_iff_le (C : G.comp_out L) (h : K ⊆ L) (D : G.comp_out K) :
+  C.hom h = D ↔ (C : set V) ⊆ (D : set V) :=
+begin
+  split,
+  { rintro rfl, exact C.sub_hom h, },
+  { revert C, refine connected_component.ind _,
+    rintro ⟨v, vL⟩ vD,
+    have h₁ : v ∈ ↑D, by
+    { refine set.mem_of_mem_of_subset _ vD,
+      simp only [set.mem_compl_iff, set_like.mem_coe, mem_supp_iff, connected_component.eq],
+      refine ⟨vL, _⟩,
+      refl, },
+    simp only [set.mem_compl_iff, mem_supp_iff, set_like.mem_coe] at h₁,
+    exact h₁.some_spec, },
+end
+
+lemma hom_refl (C : G.comp_out L) : C.hom (subset_refl L) = C :=
+by { change C.map _ = C, rw [G.out_hom_refl L, C.map_id], }
+
+lemma hom_trans (C : G.comp_out L) (h : K ⊆ L) (h' : M ⊆ K) :
+  C.hom (h'.trans h) = (C.hom h).hom h' :=
+by { change C.map _ = (C.map _).map _, rw [G.out_hom_trans, C.map_comp], }
+
+lemma hom_inf (C : G.comp_out L) (h : K ⊆ L) (Cinf : C.inf) : (C.hom h).inf :=
+set.infinite.mono (C.sub_hom h) Cinf
+
+end comp_out
+
+section ends
+
+variables (G)
+
+open category_theory
+
+-- Defined homwards for simpler use of `mathlib_fintype_inverse_systems.lean`
+instance finset_preorder_reverse : preorder (finset V) :=
+{ le := (⊇),
+  lt := (⊃),
+  le_refl := le_refl,
+  le_trans := λ a b c ab bc, by
+  { dsimp only [superset] at *, exact bc.trans ab, },
+  lt_iff_le_not_le := λ a b, by
+  { dsimp only [superset, ssuperset], exact ssubset_iff_subset_not_subset, } }
```

Comment:
Why not use `(finset V)ᵒᵈ` instead of creating a diamond?

Reply:
We had problems with `op` in when using `nonempty_sections_of_fintype_inverse_system`, and having to `unop` everywhere was just a pain. 
But maybe with more experience we can now make it work that way. Let me see.

Reply:
OK, I've pushed a version using \^o\^p and we'll see if it all goes well.

Reply:
This is not `opposite`, this is `order_dual`!

Reply:
Wait, is that just a remark on me being confused in general, or are you unhappy about the new version?

Reply:
I'm happy with the new version. In another universe, `opposite` and `order_dual` are the same type synonym anyway.

Reply:
Great! I'm a bit fearful it's going to bite us back later on but I think/guess having simple definitions is probably more important than having clean proofs.